sources/pmwebapi.go

```diff
@@ -0,0 +1,164 @@
+package sources
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type GetToken struct {
+	Context int
+}
+
+type MetricHeader struct {
+	Metrics []PcpMetrics
+}
+
+type PcpMetrics struct {
+	Name        string
+	Pmid        int64
+	Indom       int64
+	TextOneline string `json:"text-oneline"`
+	TextHelp    string `json:"text-help"`
+	Sem         string
+	Units       string
+	Type        string
+}
+
+type TimestampHeader struct {
+	Timestamp
+	Values []GetValues
+}
+
+type Timestamp struct {
+	S  string
+	Us string
+}
+
+type GetValues struct {
+	Pmid      int64
+	Name      string
+	S         string
+	Us        string
+	Instances []InstanceList
+}
+
+type InstanceList struct {
+	Instance int64
+	Value    float64
+}
+
+type pcpPmwebapiMetric struct {
+	Labels      []string
+	Labelvalues []string
+	Name        string
+	TextHelp    string
+	Value       float64
+}
+
+func init() {
+	Factories["pmwebapi"] = newPcpSource
+}
+func gaugeMetric(labels []string, labelvalues []string, name string, textHelp string, value float64) prometheus.Metric {
+	return prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "", name),
+			textHelp,
+			labels,
+			nil,
+		),
+		prometheus.GaugeValue,
+		value,
+		labelvalues...,
+	)
+}
+
+type pcpPmwebapiSource struct {
+	pcpPmwebapiMetrics []pcpPmwebapiMetric
+}
+
+func (p *pcpPmwebapiSource) Update(ch chan<- prometheus.Metric) error {
+	for _, metric := range p.pcpPmwebapiMetrics {
+		ch <- gaugeMetric(metric.Labels, metric.Labelvalues, metric.Name, metric.TextHelp, metric.Value)
+	}
+	return nil
+}
+func newPcpSource() (PcpSource, error) {
+
+	var p pcpPmwebapiSource
+
+	tokenResp, err := http.Get("http://localhost:44323/pmapi/context?hostname=localhost")
+	if err != nil {
+		log.Fatal("NewRequest: ", err)
+		return nil, err
+	}
+	defer tokenResp.Body.Close()
+
+	tokenBody, _ := ioutil.ReadAll(tokenResp.Body)
+	tokenMap := make(map[string]float64)
+	err = json.Unmarshal([]byte(tokenBody), &tokenMap)
+
+	if err != nil {
+		log.Println(err)
```

Comment:
It would be wise to do something other than just `log.Println(err)` here. The way this is currently written, if the code fails to unmarshal your `tokenBody` into `tokenMap`, it will still try and continue with the remaining code, though it will immediately fail on line 109 when it tries to access the `context` in `tokenMap`.

You can either change this to `return nil, err` or `log.Fatal(err)`. I don't have a preference, but both of these will cause your code to fail gracefully instead of something scary later on.